l2geth/core/rollup_fee.go

```diff
@@ -1,30 +1,117 @@
+	l2Fee := new(big.Int).Mul(l2GasLimit, l2GasPrice)
+	fee := new(big.Int).Add(l1Fee, l2Fee)
+	return fee.Uint64(), nil
+}
+
+// calculateL1GasLimit computes the L1 gasLimit based on the calldata and
+// constant sized overhead. The overhead can be decreased as the cost of the
+// batch submission goes down via contract optimizations.
+func calculateL1GasLimit(data []byte, overhead uint64) *big.Int {
 	zeroes, ones := zeroesAndOnes(data)
-	zeroesCost := new(big.Int).SetUint64(zeroes * params.TxDataZeroGas)
-	onesCost := new(big.Int).SetUint64((RollupBaseTxSize + ones) * params.TxDataNonZeroGasEIP2028)
-	dataCost := new(big.Int).Add(zeroesCost, onesCost)
+	zeroesCost := zeroes * params.TxDataZeroGas
+	onesCost := ones * params.TxDataNonZeroGasEIP2028
+	gasLimit := zeroesCost + onesCost + overhead
+	return new(big.Int).SetUint64(gasLimit)
+}
+
+// ceilModOneHundredMillion rounds the input integer up to the nearest modulus
+// of one hundred million
+func ceilModOneHundredMillion(num *big.Int) *big.Int {
+	if new(big.Int).Mod(num, bigHundredMillion).Cmp(common.Big0) == 0 {
+		return num
+	}
+	sum := new(big.Int).Add(num, bigHundredMillion)
+	mod := new(big.Int).Mod(num, bigHundredMillion)
+	return new(big.Int).Sub(sum, mod)
+}
+
+// VerifyL1GasPrice returns an error if the number is an invalid possible L1 gas
+// price
+func VerifyL1GasPrice(l1GasPrice *big.Int) error {
+	if new(big.Int).Mod(l1GasPrice, bigHundredMillion).Cmp(common.Big0) != 0 {
+		return errInvalidGasPrice
+	}
+	return nil
+}
 
-	// get the data fee
-	dataFee := new(big.Int).Mul(dataPrice, dataCost)
-	executionFee := new(big.Int).Mul(executionPrice, new(big.Int).SetUint64(gasUsed))
-	fee := new(big.Int).Add(dataFee, executionFee)
-	return fee
+// VerifyL2GasPrice returns an error if the number is an invalid possible L2 gas
+// price
+func VerifyL2GasPrice(l2GasPrice *big.Int) error {
+	isNonZero := l2GasPrice.Cmp(common.Big0) != 0
+	isNotModHundredMillion := new(big.Int).Mod(l2GasPrice, bigHundredMillion).Cmp(common.Big1) != 0
+	if isNonZero && isNotModHundredMillion {
+		return errInvalidGasPrice
+	}
+	if l2GasPrice.Cmp(common.Big0) == 0 {
+		return errInvalidGasPrice
+	}
+	return nil
+}
+
+// RoundL1GasPrice returns a ceilModOneHundredMillion where 0
+// is the identity function
+func RoundL1GasPrice(gasPrice *big.Int) *big.Int {
+	return ceilModOneHundredMillion(gasPrice)
+}
+
+// RoundL2GasPriceBig implements the algorithm:
+// if gasPrice is 0; return 1
+// if gasPrice is 1; return 1**9 + 1
+// return ceilModOneHundredMillion(gasPrice-1)+1
+func RoundL2GasPrice(gasPrice *big.Int) *big.Int {
+	if gasPrice.Cmp(common.Big0) == 0 {
+		return big.NewInt(1)
+	}
+	if gasPrice.Cmp(common.Big1) == 0 {
+		return new(big.Int).Add(bigHundredMillion, common.Big1)
+	}
+	gp := new(big.Int).Sub(gasPrice, common.Big1)
+	mod := ceilModOneHundredMillion(gp)
+	return new(big.Int).Add(mod, common.Big1)
+}
+
+func DecodeL2GasLimit(gasLimit uint64) uint64 {
```

Comment:
This is dead code

Reply:
This is still used but cannot be used in `types.Transaction.L2Gas` due to a cyclic dependency